### ISSUE TITLE
Demes does not support `inf` for start_size. 

### ISSUE THREAD

**Issue by georgette-femerling** [MomentsLD/moments#173]
When doing inference with demes and no upper bound is defined in the `options.yaml`, it results in an error from the Demes package. 

```
File <attrs generated init demes.demes.Epoch>:12, in __init__(self, start_time, end_time, start_size, end_size, size_function, selfing_rate, cloning_rate)
     11 __attr_validator_end_time(self, __attr_end_time, self.end_time)
---> 12 __attr_validator_start_size(self, __attr_start_size, self.start_size)
     13 __attr_validator_end_size(self, __attr_end_size, self.end_size)

TypeError: deme Deme: epoch[1]: invalid epoch
```

Noticed that the `Demes.Inference._set_up_params_and_bounds` function assigns `inf` to the upper bound when not provided:
```
        if "upper_bound" in param:
            upper_bound.append(param["upper_bound"])
        else:
            upper_bound.append(np.inf)

```

but Demes class Epoch requires start_size to be finite (https://github.com/popsim-consortium/demes-python/blob/main/demes/demes.py#L178)

```
start_size: Size = attr.ib(validator=[int_or_float, positive, finite])
```
 (https://github.com/popsim-consortium/demes-python/blob/main/demes/demes.py#L178)

**Comment by apragsdale** [comment]
Hi @georgette-femerling - you're right that a deme size cannot be infinite. By setting the default upper bound to inf means that the optimization could search parameter space in an unbounded way. But I think all the attempted values would still be finite. Are you actually finding situations where moments is trying to compute stats for a model with an infinite-sized deme?

**Comment by sgravel** [comment]
In think that the issue is that the bound is automatically set to inf when not specified, and then this causes Epoch to crash. So behaviour should be either to require a finite bound (and provide a meaningful message if not provided), or to accept infinite bounds when validating the model.

**Comment by apragsdale** [comment]
At what point in the optimization is this happening? Is the scipy optimizer testing parameters that include the upper bound (inf) in one of the iterations? I haven't seen that happen before with an optimization I didn't provide bounds for.

What I'm confused about is that we are disallowed from having an initial parameter guess that is equal to the bounds - if you try this, it should raise an error. I also didn't think that scipy would test values that were exactly equal to the bound either, but would keep the tested parameters strictly within and not equal to the bounds.

But in case this is what's happening, we could enforce the bounds on the parameter search with simple changes to these lines, by setting the `<` and `>` to `<=` and `>=`. https://github.com/MomentsLD/moments/blob/e379d9f0613a57b8d8a800321eb3a8b087f5f9b5/moments/Demes/Inference.py#L935

**Comment by georgette-femerling** [comment]
Found that the crash only happens when perturbation > 0.  

If I do provide an upper bound and set pertub = 1, there is no error but results end up being highly depending on the bound value with most Ns after a bottleneck skyrocketting close to the upper_bound. This behavior was noticeable after I set a high bound (`1e10`, `5e10`) just to make it run without error. When I set the bound to be `1e5` the inferred model made more sense.

I did some more checks with perturb=1.
When providing an upper_bound, all the initial parameters are set to be somewhat close to the upper_bound, regardless of what I have in my input model.
```
Before perturbing
p0 = [10000. 20000.  5000. 30000. 20000.]
upper_bound = [1.e+10 1.e+10 1.e+10 1.e+10 1.e+10]

After perturbing
p0 = [1.e+08 1.e+08 1.e+08 1.e+08 1.e+08]
upper_bound = [1.e+10 1.e+10 1.e+10 1.e+10 1.e+10]
```

Alternatively, when not providing an upper bound, all initial parameters are set to be `nan`:
```
Before perturbing
p0 = [10000. 20000.  5000. 30000. 20000.]
upper_bound = [inf inf inf inf inf]

After perturbing
p0 = [nan nan nan nan nan]
upper_bound = [inf inf inf inf inf]
```
I'm guessing it is the `nan` that is triggering the error in demes.
